File: Lab02/altitools.py
import math


# CSCI 1913 Lab 02 Fall 2022
# Author: Bethany Freeman

def steepest_angle(var):
    """This function computes the steepest angle in degrees for a mountain.
    These values come from either a list or a tuple."""
    mountain = convert_to_list(var)

    # this checks all changes in x's that are next to each other and keeps the largest of those changes
    hold_var = mountain[0]
    x = 1
    hold_change = 0
    while x < len(mountain):
        if math.fabs(hold_var - mountain[x]) > hold_change:
            hold_change = math.fabs(hold_var - mountain[x])

        hold_var = mountain[x]
        x += 1

    # mountain_angle_r gets the angle in radians, angle_in_degree converts that angle to degrees
    mountain_angle_r = math.atan2(hold_change, 1)
    angle_in_degree = math.degrees(mountain_angle_r)

    return angle_in_degree


def total_distance(var):
    """This function compute the total climb distance of a mountain from the given values of a tuple or list."""
    mountain = convert_to_list(var)
    distance = 0
    if len(mountain) == 1 or 0:
        distance = 0
    else:
        x = 1
        while x < len(mountain):
            distance += math.sqrt(math.pow((mountain[x - 1] - mountain[x]), 2) + 1)
            x += 1
    return distance


def longest_sequential_climb(var):
    """This function takes the distance formula and uses it to return the longest sequential climb of the mountain.
    The longest sequential climb being where you go up the mountain the greatest distance in one go"""
    mountain = convert_to_list(var)

    sequential_climb = []
    if len(mountain) == 1 or 0:
        return 0
    else:
        hold = 0
        x = 1
        while x < len(mountain):
            if mountain[x - 1] < mountain[x]:
                try:
                    sequential_climb[hold] += (math.sqrt(math.pow((mountain[x - 1] - mountain[x]), 2) + 1))
                except:
                    sequential_climb.append((math.sqrt(math.pow((mountain[x - 1] - mountain[x]), 2) + 1)))

                x += 1
            else:
                hold += 1
                x += 1

    if len(sequential_climb) == 0:
        longest_travel_distance = 0
    else:
        sequential_climb.sort()
        longest_travel_distance = sequential_climb[-1]

    return longest_travel_distance


def num_of_peaks_and_valleys(var):
    """This function counts the number of peaks and valleys in the mountain and returns them in the form of a tuple.
    with peaks presenting first and valleys presenting second."""

    mountain = convert_to_list(var)

    peaks = 0
    valleys = 0
    if len(mountain) <= 2:
        peaks = 0
        valleys = 0
    else:
        for x in range(1, len(mountain) - 1):
            if mountain[x - 1] > mountain[x] < mountain[x + 1]:
                valleys += 1
            elif mountain[x - 1] < mountain[x] > mountain[x + 1]:
                peaks += 1

    return peaks, valleys


def fill_valleys(var, min_height):
    """This function takes 2 variables, 1 of type tuple or list, and 1 of type int. These variables are used to create
    a new list of variables of a hypothetical mountain range heights should the valleys in the mountain range have
    been filled with sand"""
    mountain = convert_to_list(var)

    new_list = []
    if len(mountain) == 0:
        return new_list
    else:
        x = 0
        while x < len(mountain):
            if mountain[x] < min_height:
                new_list.append(min_height)
                x += 1
            else:
                new_list.append(mountain[x])
                x += 1
        return new_list


def convert_to_list(var):
    """this takes a variable of type tuple or list and converts it into a different list as to make
    it usable by other functions and to not change the original given variable"""

    # this converts the given tuple or the given list into a list named mountain.
    converted_list = []
    if isinstance(var, tuple) or isinstance(var, list):
        for x in var:
            converted_list.append(x)

    return converted_list

File: Lab06/Altitools.java
import java.lang.Math;

/*
 * CSCI 1913
 * Lab 06
 * Author: Bethany Freeman
 * */
public class Altitools
{

    /**
     * Computes the steepest angle (Raising or falling) in a mountain range
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @return the steepest upward/downward angle in the mountain range as a positive value measured in degrees.
     */
    public static double steepest_angle(double [] mountainRange)
    {
        double [] mountain = mountainRange;

        if(mountain.length == 0)
        {
            return 0.0;
        }
        else
        {
            double holdVar = mountain[0];
            int x = 1;
            double holdChange = 0;

            while (x < mountain.length)
            {
                if(Math.abs(holdVar - mountain[x]) > holdChange)
                {
                    holdChange = Math.abs(holdVar - mountain[x]);
                }

                holdVar = mountain[x];
                x++;
            }

            double mountainAngleR = Math.atan2(holdChange, 1);
            double angleInDegree = Math.toDegrees(mountainAngleR);

            return angleInDegree;
        }
    }

    /**
     * Get the total "travel distance" over a mountain range
     * @param mountainRange an array of doubles representing the heights of a mountain range.  Will not be modified
     * @return the travel distance -- a non-negative double.
     */
    public static double total_distance(double [] mountainRange)
    {
        double [] mountain = mountainRange;
        double distance = 0;

        if(mountain.length == 1 || mountain.length == 0)
        {
            return distance;
        }
        else
        {
            int x = 1;
            while(x < mountain.length)
            {
                distance += Math.sqrt(Math.pow((mountain[x-1] - mountain[x]), 2) + 1);
                x += 1;
            }
        }

        return distance;
    }

    /**
     * compute the longest (in terms of travel distance) sequentially climbing part of a mountain range.
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @return the latest travel distance of any sequentially climbing segment of the mountain range
     */
    public static double longest_sequential_climb(double [] mountainRange)
    {
        double [] mountain = mountainRange;
        double [] sequentialClimb = new double[]{};
        double longestTravelDistance;

        if(mountain.length == 1 || mountain.length == 0)
        {
            return 0;
        }
        else
        {
            int hold = 0;
            int x = 1;

            while(x < mountain.length)
            {
                if(mountain[x-1] < mountain[x])
                {
                    try
                    {
                        sequentialClimb[hold] += (Math.sqrt(Math.pow((mountain[x-1] - mountain[x]), 2) + 1));
                    }
                    catch(Exception e)
                    {
                        sequentialClimb = new_array(sequentialClimb,(Math.sqrt(Math.pow((mountain[x-1] - mountain[x]), 2) + 1)));
                    }

                    x++;
                }
                else
                {
                    hold++;
                    x++;
                }
            }
        }

        if(sequentialClimb.length == 0)
        {
            longestTravelDistance = 0;
        }
        else
        {
           sequentialClimb = selection_sort(sequentialClimb);
           longestTravelDistance = sequentialClimb[sequentialClimb.length - 1];
        }

        return longestTravelDistance;
    }

    /**
     * compute the number of peaks (higher than it's neighbors) and valleys (lower than it's neighbors) in a mountain
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @return a length 2 int array. position 1 is the number of peaks and position 2 the number of valleys.
     */
    public static int[] num_of_peaks_and_valleys(double [] mountainRange)
    {
        double [] mountain = mountainRange;
        int [] numPeakValleys;

        int peaks = 0;
        int valleys = 0;

        if(mountain.length <= 2)
        {
            peaks = 0;
            valleys = 0;
        }
        else
        {
            for(int x = 1; x < mountain.length - 1; x++)
            {
                if (mountain[x - 1] > mountain[x] && mountain[x] < mountain[x+1])
                {
                    valleys += 1;
                }
                else if(mountain[x - 1] < mountain[x] && mountain[x] > mountain[x+1])
                {
                    peaks += 1;
                }
            }
        }

        numPeakValleys = new int[]{peaks, valleys};

        return numPeakValleys;
    }

    /**
     * Creates a new array to indicate what would happen if we were to hypothetically fill the lowest parts of a
     * mountain range up to a fixed height
     * @param mountainRange an array of doubles representing the heights of a mountain range. Will not be modified
     * @param target a double. the new array will have all positions below this filled up to this height
     * @return a new array showing the result of filling in the array
     */
    public static double[] fill_valleys(double [] mountainRange, double target)
    {
        double [] mountain = mountainRange;
        double [] newList = new double[]{};

        if(mountain.length == 0)
        {
            return newList;
        }
        else
        {
            int x = 0;
            while(x < mountain.length)
            {
                if(mountain[x] < target)
                {
                    newList = new_array(newList, target);
                    x += 1;
                }
                else
                {
                    newList = new_array(newList, mountain[x]);
                    x += 1;
                }
            }

            return newList;
        }
    }

    /**
     * Creates a new array from the passedArray and adds a value to it and returns a new array with that new value added
     * @param passedArray an array of doubles representing the current state of the array when it was given to the
     *                    function
     * @param value the value to be added to the array
     * @return a new array with the @param value added to it.
     */
    public static double[] new_array(double [] passedArray, double value)
    {
        double [] tempArray = new double[passedArray.length + 1];

       if(passedArray.length != 0)
       {
           for(int x = 0;x != passedArray.length; x++)
           {
               tempArray[x] = passedArray[x];
           }
       }

       tempArray[passedArray.length] = value;



        return tempArray;
    }

    /**
     * Based off of Zybook section 8.2.1 Selection sort. This takes the passed array and sorts it by repeatedly comparing
     * two indexes to each other until it find the smallest one and moves that one to the front, and so on until it has
     * sorted the entire array.
     * @param passedArray this is an array of double that is unsorted
     * @return a new sorted array of the passedArray
     */
    public static double[] selection_sort(double [] passedArray)
    {
        double[] tempArray = passedArray;
        double temp;

        for(int i = 0; i != tempArray.length; i++)
        {
            int smallestIndex = i;
            for(int j = i +1; j != tempArray.length; j++)
            {
                if(tempArray[j] < tempArray[smallestIndex])
                {
                    smallestIndex = j;
                }
            }

            temp = tempArray[i];
            tempArray[i] = tempArray[smallestIndex];
            tempArray[smallestIndex] = temp;
        }

        return tempArray;
    }
}




